File: Solutions/543_Diameter_of_Binary_Tree/543_Diameter_of_Binary_Tree.py
#Fourth fastest
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 1
        def depth(node):
            if not node: return 0
            L = depth(node.left)
            R = depth(node.right)
            self.ans = max(self.ans, L+R+1)
            return max(L, R) + 1

        depth(root)
        return self.ans - 1

#Third fastest
class Solution:
    def _height(self, root: TreeNode) -> int:
        if root is None:
            return 0
        return max(self._height(root.left), self._height(root.right)) + 1
    
    def _diameterOfBinaryTree(self, root: TreeNode) -> int:
        if root is None:
            return 0
        
        diameter = self._height(root.left) + self._height(root.right)
        
        return max(diameter, self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right))
    
    
    def _diameterOfBinaryTreeHelper(self, root: TreeNode) -> (int, int):
        if root is None:
            return 0, 0
        
        right_height, right_diameter = self._diameterOfBinaryTreeHelper(root.right)
        left_height, left_diameter = self._diameterOfBinaryTreeHelper(root.left)
        
        return max(left_height, right_height) + 1, max(left_height + right_height, left_diameter, right_diameter)
    
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        return self._diameterOfBinaryTreeHelper(root)[-1]

#Second fastest
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        res = 0
        
        def helper(root):
            nonlocal res
            if not root: return 0
            left = helper(root.left)
            right = helper(root.right)
            res = max(res, left + right)
            return max(left, right) + 1
        
        helper(root)
        
        return res

#Fastest
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        
        self.res = 1
        def Dimeter(root):
            if not root:
                return 0

            left = Dimeter(root.left)
            right = Dimeter(root.right)

            temp = 1 + max(left, right)
            self.res = max(self.res, 1 + left + right)

            return temp
        
        Dimeter(root)
        return self.res - 1

File: Solutions/543_Diameter_of_Binary_Tree/543_Diameter_of_Binary_Tree.js
var diameterOfBinaryTree = function(root) {
    let diameter = 0
    if(!root) return 0
    function computeDiameter(root) {
        if(!root) {
            return 0
        }
        const d1 = computeDiameter(root.left)
        const d2 = computeDiameter(root.right)
        diameter = Math.max(diameter, d1 + d2)
        return Math.max(d1, d2) + 1
    }
    computeDiameter(root)
    return diameter
};

//Third fastest
let diameterOfBinaryTree = function(root) {
    let max = 0;

    // assume node !== null
    let postOrderDFVisit = function(node) {
        let leftHeight = 0, rightHeight = 0;
        if (node.left) {
            postOrderDFVisit(node.left);
            leftHeight = node.left.height;
        }
        if (node.right) {
            postOrderDFVisit(node.right);
            rightHeight = node.right.height;
        }

        node.height = Math.max(1 + leftHeight, 1 + rightHeight);
        node.diameter = leftHeight + rightHeight;
        if (node.diameter > max) {
            max = node.diameter;
        }
    };

    if (!root) {
        return 0;
    }

    postOrderDFVisit(root);

    return max;
};

//Second fastest
var diameterOfBinaryTree = function(root) {
    let max = 0;
    
    const dfs = (node, level) => {
        if (!node) {
            return 0;
        }
        
        const left = dfs(node.left);
        const right = dfs(node.right);
        
        max = Math.max(max, left + right);
        return Math.max(left, right) + 1;
    }
    
    dfs(root, 0);
    
    return max;
    
};

//Fastest
const diameterOfBinaryTree = root => {
    let max = 0
    const traverse = node => {
        if (!node) return 0
        const left = traverse(node.left)
        const right = traverse(node.right)
        max = Math.max(left + right, max)
        return Math.max(left, right) + 1 
    }
    traverse(root)
    return max
}